WordPress/gutenberg
# Expanded `wp-env` Lifecycle Scripts

## What?
In https://github.com/WordPress/gutenberg/pull/50196 I added an `afterSetup` script that is executed after `wp-env start` and `wp-env clean`.  This pull request expands on that work and provides more granular lifecycle scripts in its place.

## Why?
Once I began adopting the `afterSetup` feature I found that it wasn't granular enough and that there were cases where it would also be useful to have a script execute.

## How?
We're going to remove the `afterSetup` option in favor of a top-level `lifecycleScripts` option with entries for each lifecycle script that the user wants to implement. These scripts can all be overridden using `WP_ENV_LIFECYCLE_SCRIPT_{ EVENT_NAME }` environment variables. For now this includes `afterStart`, `afterClean`, and `beforeDestroy`. I've also made the execution of the script asynchronous because it was blocking the spinner.

## Testing Instructions

1. Make sure `start`, `clean`, and `destroy` all work the same without any scripts set.
2. Create `.wp-env.override.json` with content:
```
{
    "lifecycleScripts": {
        "afterStart": "echo 'Test Start' && sleep 10 && echo 'OK'",
        "afterClean": "echo 'Test Clean' && sleep 10 && echo 'OK'",
        "beforeDestroy": "echo 'Test Destroy' && sleep 10 && echo 'OK'"
    }
}
```
3. Check `start`, `clean`, and `destroy` and watch the output.
4. You can run the above with `--debug` and it will persist the command output.
5. Run `export WP_ENV_LIFECYCLE_SCRIPT_AFTER_START="echo 'Test Override' && sleep 10"` and run `start`. Note that it is overridden.
6. Run `export WP_ENV_LIFECYCLE_SCRIPT_AFTER_START="echo "FAIL" 1>&2 && false"` and run `start`. It should give the error and indicate the failure.
